### PR TITLE
feat(persistent_session): Make schema more flexible

### PR DESCRIPTION
**Pull request by ieQu1**

### CODE REVIEW

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 1]
```diff
@@ -509,22 +509,39 @@ emqx_schema {
...
+        zh: "用于存储持久性会话信息的数据库后端\n"
+            "- `builtin`: 使用一个嵌入式数据库（mria）"
+    }
+    label: {
+        en: "Backend"
+        zh: "后台"
+    }
+  }
+
+
```
Comment (ieQu1) [inline apps/emqx/i18n/emqx_schema_i18n.conf:525]: Extra \n

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 2]
```diff
@@ -509,22 +509,39 @@ emqx_schema {
...
-            "群集重新启动，代价是磁盘使用量增加，吞吐量降低。\n"
-    }
-    label: {
-        en: "Storage type"
-        zh: "存储类型"
+  persistent_session_store_backend {
+    desc {
+        en: "Database backend used to store information about the persistent sessions.\n"
+            "- `builtin`: Use an embedded database (mria)"
+        zh: "用于存储持久性会话信息的数据库后端\n"
```
Comment (ieQu1) [inline apps/emqx/i18n/emqx_schema_i18n.conf:516]: Warning: machine translation

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 3]
```diff
@@ -509,22 +509,38 @@ emqx_schema {
...
-            "cluster restart, at the price of more disc usage and less throughput.\n"
-        zh: "将有关持久会话的信息存储在磁盘或内存中。\n"
-            "如果选择了ram，有关持久会话的所有信息将保留\n"
-            "只要群集中至少有一个节点处于活动状态，就可以保留信息。\n"
-            "如果选择了光盘，则信息将保留在光盘上，并且将继续存在\n"
-            "群集重新启动，代价是磁盘使用量增加，吞吐量降低。\n"
+        en: "Database backend used to store information about the persistent sessions.\n"
+            "- `builtin`: Use an embedded database (mria)"
+        zh: "用于存储持久性会话信息的数据库后端\n"
+            "- `builtin`: 使用一个嵌入式数据库（mria）"
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:517]:
```suggestion
            "- `builtin`: 使用一个内置数据库（mria）"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 4]
```diff
@@ -509,22 +509,38 @@ emqx_schema {
...
     }
     label: {
-        en: "Storage type"
-        zh: "存储类型"
+        en: "Backend"
+        zh: "后台"
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:521]:
```suggestion
        zh: "后端类型"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 5]
```diff
@@ -509,22 +509,38 @@ emqx_schema {
...
+  }
+
+  persistent_store_on_disc {
+    desc {
+        en: "Persist session data on disc. If `false`, the data will be stored in RAM."
+        zh: "将会话数据保留在磁盘上。如果`false`，数据将被存储在RAM中。"
+    }
+    label: {
+        en: "Persist on disc"
+        zh: "坚持 在光盘上"
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:532]:
```suggestion
        zh: "持久化在磁盘上"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 6]
```diff
@@ -509,22 +509,38 @@ emqx_schema {
...
-        zh: "存储类型"
+        en: "Backend"
+        zh: "后台"
+    }
+  }
+
+  persistent_store_on_disc {
+    desc {
+        en: "Persist session data on disc. If `false`, the data will be stored in RAM."
+        zh: "将会话数据保留在磁盘上。如果`false`，数据将被存储在RAM中。"
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:528]:
```suggestion
        zh: "将会话数据保存在磁盘上。如果为 `false` 则存储在内存中。"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 7]
```diff
@@ -509,22 +509,38 @@ emqx_schema {
...
+    label: {
+        en: "Persist on disc"
+        zh: "坚持 在光盘上"
+    }
+  }
+
+  persistent_store_ram_cache {
+    desc {
+        en: "Maintain a copy of the data in RAM for faster access."
+        zh: "在RAM中保持一份数据的副本，以便更快地访问。"
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:539]:
```suggestion
        zh: "在内存中保持一份数据的副本，以便更快地访问。"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 8]
```diff
@@ -509,22 +509,38 @@ emqx_schema {
...
+    }
+    label: {
+        en: "RAM cache"
+        zh: "RAM缓存"
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:543]:
```suggestion
        zh: "内存缓存"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 9]
```diff
@@ -569,6 +585,39 @@ emqx_schema {
     }
   }
 
+  persistent_session_builtin_session_table {
+    desc {
+         en: "Tuning for built-in session table."
+         zh: "对内置的会话表进行调整。"
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:591]:
```suggestion
         zh: "优化内置会话表的配置。"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 10]
```diff
@@ -569,6 +585,39 @@ emqx_schema {
...
+    }
+    label: {
+         en: "Persistent session."
+         zh: "持久会话。"
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:595]:
```suggestion
         zh: "持久会话"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 11]
```diff
@@ -569,6 +585,39 @@ emqx_schema {
     }
   }
 
+  persistent_session_builtin_session_table {
+    desc {
+         en: "Tuning for built-in session table."
+         zh: "对内置的会话表进行调整。"
+    }
+    label: {
+         en: "Persistent session."
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:594]: If I remember correctly, the Label should not contain punctuation
```suggestion
         en: "Persistent session"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 12]
```diff
@@ -569,6 +585,39 @@ emqx_schema {
...
+         zh: "持久会话。"
+    }
+  }
+
+  persistent_session_builtin_sess_msg_table {
+    desc {
+         en: "Tuning for built-in session messages table."
+         zh: "对内置的会话信息表进行调整。"
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:602]:
```suggestion
         zh: "优化内置的会话消息表的配置。"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 13]
```diff
@@ -569,6 +585,39 @@ emqx_schema {
...
+    }
+    label: {
+         en: "Persistent session messages."
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:605]:
```suggestion
         en: "Persistent session messages"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 14]
```diff
@@ -569,6 +585,39 @@ emqx_schema {
...
+         zh: "持久的会话信息。"
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:606]:
```suggestion
         zh: "持久化会话消息"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 15]
```diff
@@ -569,6 +585,39 @@ emqx_schema {
...
+    }
+  }
+
+  persistent_session_builtin_messages_table {
+    desc {
+         en: "Tuning for built-in messages table."
+         zh: "对内置信息表进行调校。"
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:613]:
```suggestion
         zh: "优化内置消息表配置。"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 16]
```diff
@@ -569,6 +585,39 @@ emqx_schema {
...
+    }
+    label: {
+         en: "Persistent messages."
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:616]:
```suggestion
         en: "Persistent messages"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 17]
```diff
@@ -569,6 +585,39 @@ emqx_schema {
...
+         zh: "持久性信息。"
```
Comment (HJianBo) [inline apps/emqx/i18n/emqx_schema_i18n.conf:617]:
```suggestion
         zh: "持久化消息"
```

File: apps/emqx/i18n/emqx_schema_i18n.conf (hunk 2 above)
Comment (HJianBo) [reply apps/emqx/i18n/emqx_schema_i18n.conf:516]: I modified parts of it to be it looks more localized (maybe). it looks like each one configuration needs a lot of explanation if it needs to be clearly understood. 😭

File: apps/emqx/test/emqx_persistent_session_SUITE.erl [hunk 18]
```diff
@@ -193,7 +193,9 @@ end_per_group(gc_tests, Config) ->
     meck:unload(mnesia),
     ?config(store_owner, Config) ! stop,
     ok;
-end_per_group(Group, _Config) when Group =:= ram_tables; Group =:= disc_tables ->
+end_per_group(Group, _Config) when
+    Group =:= ram_tables; Group =:= disc_tables; Group =:= rocks_tables
```
Comment (qzhuyan) [inline apps/emqx/test/emqx_persistent_session_SUITE.erl:197]: rocksdb_tables ?

File: apps/emqx/test/emqx_persistent_session_SUITE.erl [hunk 19]
```diff
@@ -69,6 +69,12 @@ groups() ->
             {group, snabbkaffe},
             {group, gc_tests}
         ]},
+        {rocks_tables, [], [
```
Comment (qzhuyan) [inline apps/emqx/test/emqx_persistent_session_SUITE.erl:72]: rocksdb_tables?

**Review by qzhuyan — APPROVED** [review]

File: apps/emqx/i18n/emqx_schema_i18n.conf (hunk 2 above)
Comment (ieQu1) [reply apps/emqx/i18n/emqx_schema_i18n.conf:516]: Thanks for the comment, I will make descriptions more in-depth.

File: apps/emqx/test/emqx_persistent_session_SUITE.erl (hunk 19 above)
Comment (ieQu1) [reply apps/emqx/test/emqx_persistent_session_SUITE.erl]: I removed this part, I didn't go with that idea.

File: apps/emqx/i18n/emqx_schema_i18n.conf (hunk 2 above)
Comment (ieQu1) [reply apps/emqx/i18n/emqx_schema_i18n.conf:516]: Does it make more sense now https://github.com/emqx/emqx/pull/7897/commits/0f8829dea1e929e3d1d6439a7d1d80a8ffc9639b ?

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 20]
```diff
@@ -569,6 +587,39 @@ emqx_schema {
...
+    }
+    label: {
+         en: "Persistent session."
+         zh: "持久会话。"
+    }
+  }
+
+  persistent_session_builtin_sess_msg_table {
+    desc {
+         en: "Performance tuning options for built-in session messages table."
```
Comment (thalesmg) [inline apps/emqx/i18n/emqx_schema_i18n.conf:603]: I'm trying to remember: are the "session messages" pointers from (possibly) different persistent sessions to the message payload itself (stored in the "messages" table)?

Comment (ieQu1) [reply apps/emqx/i18n/emqx_schema_i18n.conf:603]: Yes.

File: apps/emqx/i18n/emqx_schema_i18n.conf (hunk 9 above)
Comment (ieQu1) [reply apps/emqx/i18n/emqx_schema_i18n.conf:593]: Thanks, I will add that to the batch

**Review by thalesmg — APPROVED** [review]

File: apps/emqx/src/persistent_session/emqx_persistent_session_backend_builtin.erl [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+    put_message/1,
+    get_message/1,
+    ro_transaction/1
+]).
+
+-type mria_table_type() :: ram_copies | disc_copies | rocksdb_copies.
+
+-define(IS_ETS(BACKEND), (BACKEND =:= ram_copies orelse BACKEND =:= disc_copies)).
+
+create_tables() ->
```
[the remark this answers is not in the document]
Comment (ieQu1) [reply apps/emqx/src/persistent_session/emqx_persistent_session_backend_builtin.erl:44]: Good point. At the time of implementation this feature wasn't available.

[the remark this answers is not in the document]
Comment (ieQu1) [reply apps/emqx/src/persistent_session/emqx_persistent_session_backend_builtin.erl:44]: I will fix this in the next PR, that will be mostly focused on performance tuning.

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 22]
```diff
@@ -509,22 +509,42 @@ emqx_schema {
...
-            "If disc is chosen, the information is persisted on disc and will survive\n"
-            "cluster restart, at the price of more disc usage and less throughput.\n"
-        zh: "将有关持久会话的信息存储在磁盘或内存中。\n"
-            "如果选择了ram，有关持久会话的所有信息将保留\n"
-            "只要群集中至少有一个节点处于活动状态，就可以保留信息。\n"
-            "如果选择了光盘，则信息将保留在光盘上，并且将继续存在\n"
-            "群集重新启动，代价是磁盘使用量增加，吞吐量降低。\n"
+        en: "Database management system used to store information about persistent sessions and messages.\n"
+            "- `builtin`: Use the embedded database (mria)"
+        zh: "数据库管理系统，用于存储有关持久性会话和消息的信息\n"
```
Comment (qzhuyan) [inline apps/emqx/i18n/emqx_schema_i18n.conf:516]: Database management system = "数据库管理系统", sounds more like a DB frontend (view records, change data on the fly)  in Chinese.
To me, `Database` should be fine.

**Review by thalesmg — DISMISSED** [review]

**Review by HJianBo — APPROVED** [review]
Desc is ok for me. 
Hint: rebase the master branch could resolve failed gateway cases

**Pull request merged**